# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1688584361 -7200
# Parent  48c8095010bf93e173902ef8cdbe08ba769eee87
Bug 1841348 - remove dysfunctional fallbacks to LoginManagerAuthPrompter. r=leftmostcat

Services.ww.getNewAuthPrompter() returns LoginManagerAuthPrompter which implements nsIAuthPrompt
but no longer implements nsIAuthPrompt.promptPassword() and nsIAuthPrompt.promptUsernameAndPassword().

So remove any fallbacks from MsgAuthPrompt to those methods.

diff --git a/calendar/base/content/publish.js b/calendar/base/content/publish.js
--- a/calendar/base/content/publish.js
+++ b/calendar/base/content/publish.js
@@ -181,17 +181,19 @@ function publishItemArray(aItemArray, aP
   }
 }
 
 var notificationCallbacks = {
   // nsIInterfaceRequestor interface
   getInterface(iid, instance) {
     if (iid.equals(Ci.nsIAuthPrompt)) {
       // use the window watcher service to get a nsIAuthPrompt impl
-      return Services.ww.getNewAuthPrompter(null);
+      return Cc["@mozilla.org/messenger/msgAuthPrompt;1"].getService(
+        Ci.nsIAuthPrompt
+      );
     }
 
     throw Components.Exception("", Cr.NS_ERROR_NO_INTERFACE);
   },
 };
 
 /**
  * Listener object to pass to `channel.asyncOpen()`. A reference to the current dialog window
diff --git a/mailnews/base/src/MsgIncomingServer.jsm b/mailnews/base/src/MsgIncomingServer.jsm
--- a/mailnews/base/src/MsgIncomingServer.jsm
+++ b/mailnews/base/src/MsgIncomingServer.jsm
@@ -835,18 +835,20 @@ class MsgIncomingServer {
     let ok;
     let authPrompt;
     try {
       // This prompt has a checkbox for saving password.
       authPrompt = Cc["@mozilla.org/messenger/msgAuthPrompt;1"].getService(
         Ci.nsIAuthPrompt
       );
     } catch (e) {
-      // Often happens in tests. This prompt has no checkbox for saving password.
-      authPrompt = Services.ww.getNewAuthPrompter(null);
+      throw Components.Exception(
+        "Can't prompt for password",
+        Cr.NS_ERROR_ABORT
+      );
     }
     if (this.username) {
       ok = authPrompt.promptPassword(
         promptTitle,
         promptMessage,
         this.serverURI,
         Ci.nsIAuthPrompt.SAVE_PASSWORD_PERMANENTLY,
         outPassword
diff --git a/mailnews/compose/src/SmtpServer.jsm b/mailnews/compose/src/SmtpServer.jsm
--- a/mailnews/compose/src/SmtpServer.jsm
+++ b/mailnews/compose/src/SmtpServer.jsm
@@ -218,18 +218,20 @@ class SmtpServer {
   getPasswordWithUI(promptMessage, promptTitle) {
     let authPrompt;
     try {
       // This prompt has a checkbox for saving password.
       authPrompt = Cc["@mozilla.org/messenger/msgAuthPrompt;1"].getService(
         Ci.nsIAuthPrompt
       );
     } catch (e) {
-      // Often happens in tests. This prompt has no checkbox for saving password.
-      authPrompt = Services.ww.getNewAuthPrompter(null);
+      throw Components.Exception(
+        "Can't prompt for password",
+        Cr.NS_ERROR_ABORT
+      );
     }
     let password = this._getPasswordWithoutUI();
     if (password) {
       this.password = password;
       return this.password;
     }
     let outUsername = {};
     let outPassword = {};
diff --git a/mailnews/news/src/nsNewsFolder.cpp b/mailnews/news/src/nsNewsFolder.cpp
--- a/mailnews/news/src/nsNewsFolder.cpp
+++ b/mailnews/news/src/nsNewsFolder.cpp
@@ -23,17 +23,16 @@
 #include "nsNewsUtils.h"
 
 #include "nsIMsgIncomingServer.h"
 #include "nsINntpIncomingServer.h"
 #include "nsINewsDatabase.h"
 #include "nsILineInputStream.h"
 
 #include "nsIMsgWindow.h"
-#include "nsIWindowWatcher.h"
 
 #include "nsNetUtil.h"
 #include "nsIAuthPrompt.h"
 #include "nsIURL.h"
 #include "nsNetCID.h"
 #include "nsINntpUrl.h"
 
 #include "nsNewsDownloader.h"
@@ -1059,20 +1058,17 @@ nsMsgNewsFolder::GetAuthenticationCreden
   }
 
   // Show the prompt if we need to
   if (mustPrompt ||
       (mayPrompt && (mGroupUsername.IsEmpty() || mGroupPassword.IsEmpty()))) {
     nsCOMPtr<nsIAuthPrompt> authPrompt =
         do_GetService("@mozilla.org/messenger/msgAuthPrompt;1");
     if (!authPrompt) {
-      nsCOMPtr<nsIWindowWatcher> wwatch(
-          do_GetService(NS_WINDOWWATCHER_CONTRACTID));
-      if (wwatch) wwatch->GetNewAuthPrompter(0, getter_AddRefs(authPrompt));
-      if (!authPrompt) return NS_ERROR_FAILURE;
+      return NS_ERROR_FAILURE;
     }
 
     if (authPrompt) {
       // Format the prompt text strings
       nsString promptTitle, promptText;
       bundle->GetStringFromName("enterUserPassTitle", promptTitle);
 
       nsString serverName;
